### PR TITLE
feat: Implement external surface navigation

### PR DESCRIPTION
**Pull request by Corentin-Allaire**
Implement external surface in the navigator state as multimap of layer index and GeometryIdentifier. We will try to visit all the surfaces in the map by ignoring the boundary check for those surfaces but if multiple surface are on the same plane that might not be possible. 

The KF have been updated so that all the measurement surface gets added to the external surfaces.

### CODE REVIEW

**Comment by Corentin-Allaire** [comment]
@asalzburger since we discuss this two week ago can I ask you to review this PR and tell me if this correspond to what you where thinking or if I should change something ?

**Comment by asalzburger** [comment]
Sure, I had that on my watch list, I will have a look today!

File: Core/include/Acts/Geometry/detail/Layer.ipp [hunk 1]
```diff
@@ -143,9 +143,15 @@ std::vector<SurfaceIntersection> Layer::compatibleSurfaces(
     if (!acceptSurface(sf, sensitive)) {
       return;
     }
+    bool boundaryCheck = options.boundaryCheck;
+    if (std::find(options.externalSurfaces.begin(),
+                  options.externalSurfaces.end(),
+                  sf.geometryId()) != options.externalSurfaces.end()) {
+      boundaryCheck = false;
+    }
```
Comment (asalzburger) [inline Core/include/Acts/Geometry/detail/Layer.ipp:151]: Do you think we'd need a `options.boundaryCheckNavigation` and one for `options.boundaryCheckExternal` ?
I am not sure.

**Review by asalzburger — APPROVED** [review]
Looks good, I have only one question on the boundary checks.

**Comment by asalzburger** [comment]
Did you manage to test/validate that it's doing what it is supposed to?

**Comment by Corentin-Allaire** [comment]
> Did you manage to test/validate that it's doing what it is supposed to?

Yes I tested it by adding "manually" some external surface to the navigator and using some cout. I could have a look to add some test in the unit. The only "issue" is that this won't work if you have external surface on the same plane (same R in the barrel or same Z en the endcap).

File: Core/include/Acts/Geometry/detail/Layer.ipp (hunk 1 above)
Comment (Corentin-Allaire) [reply Core/include/Acts/Geometry/detail/Layer.ipp:151]: I don't think that would be needed. If you had 'options.boundaryCheckExternal=true' that be the same thing as not having external surface (you can just keep 'options.externalSurfaces={}'). I wouldn't cost anything to have an extra option but I don't think this is needed.

**Comment by asalzburger** [comment]
> > Did you manage to test/validate that it's doing what it is supposed to?
> 
> Yes I tested it by adding "manually" some external surface to the navigator and using some cout. I could have a look to add some test in the unit. The only "issue" is that this won't work if you have external surface on the same plane (same R in the barrel or same Z en the endcap).

Yes, probably we should add a unit test for this, anyhow, I guess for the moment I can merge this in.

**Comment by Corentin-Allaire** [comment]
> > > Did you manage to test/validate that it's doing what it is supposed to?
> > 
> > 
> > Yes I tested it by adding "manually" some external surface to the navigator and using some cout. I could have a look to add some test in the unit. The only "issue" is that this won't work if you have external surface on the same plane (same R in the barrel or same Z en the endcap).
> 
> Yes, probably we should add a unit test for this, anyhow, I guess for the moment I can merge this in.

I will have a look at that and tomorrow I should be able to test this in athena.

**Pull request merged**